simulate/simulate.cc

```diff
@@ -2632,18 +2632,42 @@ void Simulate::Render() {
   }
 
   // user figures
+  if (this->newfigurerequest.load() == 1) {
+    this->user_figures_.clear();
+    for (const auto& [viewport, figure] : this->user_figures_new_) {
+      this->user_figures_.push_back(std::make_pair(viewport, figure));
+    }
+    int value = 1;
+    this->newfigurerequest.compare_exchange_strong(value, 0);
+  }
   for (auto& [viewport, figure] : this->user_figures_) {
     ShowFigure(this, viewport, &figure);
   }
 
   // overlay text
+  if (this->newtextrequest.load() == 1) {
+    this->user_texts_.clear();
+    for (const auto& [font, gridpos, text1, text2] : this->user_texts_new_) {
+      this->user_texts_.push_back(std::make_tuple(font, gridpos, text1, text2));
+    }
+    int value = 1;
+    this->newtextrequest.compare_exchange_strong(value, 0);
+  }
   for (auto& [font, gridpos, text1, text2] : this->user_texts_) {
     ShowOverlayText(this, rect, font, gridpos, text1, text2);
   }
 
   // user images
+  if (this->newimagerequest.load() == 1) {
```

Comment:
Make sure that if you use shared_ptr, user_images_new_ gets cleared so that the images aren't held in memory unnecessarily. I think if you use std::swap you could switch to unique_ptr.